### PR TITLE
feat: FSDP2 SFT

### PR DESCRIPTION
**Pull request by yfw**
# What does this PR do ?

Adds changes for SFT w/ FSDP2

# Issues
Closes https://github.com/NVIDIA/reinforcer/issues/197

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA/reinforcer/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/reinforcer/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/reinforcer/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
Test runs:

TP=1, 1 node, 2 gpus
https://wandb.ai/nvidia/sft-dev-yifu/runs/yp6769f1/overview
`uv run examples/run_sft.py logger.wandb_enabled=True cluster.gpus_per_node=2 policy.precision=bfloat16 policy.model_name=meta-llama/Llama-3.1-8B-Instruct policy.dtensor_cfg.enabled=True logger.wandb.name=sft-fsdp2-tp1-llama3.1-8b-seed42-fix sft.seed=42
`

TP=2, 1 node, 2 gpus
https://wandb.ai/nvidia/sft-dev-yifu/runs/gspnpvrx/overview
`uv run examples/run_sft.py logger.wandb_enabled=True cluster.gpus_per_node=2 policy.precision=bfloat16 policy.model_name=meta-llama/Llama-3.1-8B-Instruct policy.dtensor_cfg.enabled=True policy.dtensor_cfg.tensor_parallel_size=2 logger.wandb.name=sft-fsdp2-tp2-llama3.1-8b-seed42-fix sft.seed=42
`

TP=2 w/ SP, 1 node, 2 gpus
https://wandb.ai/nvidia/sft-dev-yifu/runs/te65amfn/overview
`uv run examples/run_sft.py logger.wandb_enabled=True cluster.gpus_per_node=2 policy.precision=bfloat16 policy.model_name=meta-llama/Llama-3.1-8B-Instruct policy.dtensor_cfg.enabled=True policy.dtensor_cfg.tensor_parallel_size=2 policy.dtensor_cfg.sequence_parallel=True logger.wandb.name=sft-fsdp2-tp2-sp-llama3.1-8b-seed42 sft.seed=42
`

TP=2 w/ activation checkpointing, 1 node, 2 gpus
https://wandb.ai/nvidia/sft-dev-yifu/runs/1mremwtx/overview
`uv run examples/run_sft.py sft.val_at_start=True checkpointing.enabled=False logger.wandb_enabled=True logger.wandb.project=sft-dev-yifu cluster.gpus_per_node=2 policy.precision=bfloat16 policy.model_name=meta-llama/Llama-3.1-8B-Instruct policy.dtensor_cfg.enabled=True policy.dtensor_cfg.tensor_parallel_size=2 policy.dtensor_cfg.activation_checkpointing=True logger.wandb.name=sft-fsdp2-tp2-ac-llama3.1-8b-seed42 sft.seed=42
`

TP=8 w/ SP, 4 nodes, 32 gpus, 32k context length
https://wandb.ai/nvidia/sft-dev-yifu/runs/95a9h3cd/overview
`uv run examples/run_sft.py logger.wandb_enabled=True cluster.gpus_per_node=8 cluster.num_nodes=4 policy.precision=bfloat16 policy.model_name=meta-llama/Llama-3.1-8B-Instruct policy.dtensor_cfg.enabled=True policy.dtensor_cfg.tensor_parallel_size=8 policy.dtensor_cfg.sequence_parallel=True policy.max_total_sequence_length=32000 logger.wandb.name=sft-fsdp2-8node-32kcontext-tp8-sp-llama3.1-8b-seed42 sft.seed=42
`

TP=8 w/ SP and activation checkpointing, 4 nodes, 32 gpus, Qwen2.5 32B, 16k context length
https://wandb.ai/nvidia/sft-dev-yifu/runs/e5mi4wko
`uv run examples/run_sft.py logger.wandb_enabled=True  cluster.gpus_per_node=8 cluster.num_nodes=4 policy.precision=bfloat16 policy.model_name=Qwen/Qwen2.5-32B-Instruct policy.dtensor_cfg.enabled=True policy.dtensor_cfg.tensor_parallel_size=8 policy.dtensor_cfg.sequence_parallel=True policy.dtensor_cfg.activation_checkpointing=True policy.max_total_sequence_length=16000 logger.wandb.name=sft-fsdp2-4node-16kcontext-tp8-sp-ac-qwen2.5-32b-seed42 sft.seed=42`

### CODE REVIEW

File: nemo_reinforcer/models/policy/dtensor_policy_worker.py [hunk 1]
```diff
@@ -261,114 +261,121 @@ def train(
...
-
-        # Get data from batch and move to device
-        data.to("cuda")
-
-        losses = []
-        all_mb_metrics = []
-        for gb_start in range(0, dataset_size, local_gbs):
-            self.optimizer.zero_grad()
-            mb_losses = []
+        if eval_mode:
```
Comment (yfw) [inline nemo_reinforcer/models/policy/dtensor_policy_worker.py]: Can ignore the diffs here, I just added torch.no_grad() when in eval mode, so diffs are due to spacing.

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 2]
```diff
@@ -168,14 +168,22 @@ def __call__(
...
         # Gather the logprobs for the actual next tokens
-        token_logprobs = logprobs.gather(
-            dim=-1, index=next_tokens.unsqueeze(-1)
-        ).squeeze(-1)
+        if isinstance(next_token_logits, torch.distributed.tensor.DTensor):
+            token_logprobs = get_logprobs_from_vocab_parallel_logits(
+                next_token_logits, data["input_ids"]
+            )
+        else:
+            next_tokens = data.get("input_ids")[:, 1:].cuda()  # Skip first token
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/loss_functions.py:179]: nit: can we make ClippedPGLossFn https://github.com/NVIDIA/reinforcer/blob/98b7a90c41fc1f86f2d90f49e2f09c726f417ce5/nemo_reinforcer/algorithms/loss_functions.py#L115 match this?

File: nemo_reinforcer/models/policy/dtensor_policy_worker.py [hunk 3]
```diff
@@ -348,7 +349,7 @@ def train(
             local_loss = torch.tensor(losses, device="cuda")
             global_loss = torch.zeros_like(local_loss)
             torch.distributed.all_reduce(local_loss)
-            global_loss = local_loss / self.dp_size
+            global_loss = local_loss / (self.dp_size * self.tp_size)
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/dtensor_policy_worker.py]: is this b/c all ranks participate? should we just be reducing over `self.dp_mesh.get_group`?

File: nemo_reinforcer/algorithms/loss_functions.py (hunk 2 above)
Comment (yfw) [reply nemo_reinforcer/algorithms/loss_functions.py:179]: Updated https://github.com/NVIDIA/reinforcer/pull/206/commits/4f5fac44839104d874791f3f78acabdb0e20949f#diff-a6b9b52523da66e6629ac1b5cea3d6fc490c6f3bb3970285c78820e98990bcb7L115

File: nemo_reinforcer/models/policy/dtensor_policy_worker.py (hunk 3 above)
Comment (yfw) [reply nemo_reinforcer/models/policy/dtensor_policy_worker.py]: Yes, good point. Updated here https://github.com/NVIDIA/reinforcer/pull/206/commits/4f5fac44839104d874791f3f78acabdb0e20949f#diff-59218d9231a3261f1e26a7cfade3f890415ac2f37bae3b23bbdf055322bb84f8L351
